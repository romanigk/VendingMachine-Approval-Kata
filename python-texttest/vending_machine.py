
class VendingMachine:
    def __init__(self):
        self.display = ""
        self.coins = []
        self.returns = []
        self.balance = 0
        self.accepted_coins = [5,10,25]
        self._display_balance()

    def _display_balance(self):
        if self.balance:
            self.display = f"{self.balance}"
        else:
            self.display = "INSERT COIN"

    def insert_coin(self, value):
        if value in self.accepted_coins:
            self.coins.append(value)
            self.balance += value
            self._display_balance()
        else:
            self.returns.append(value)

    def return_coins(self):
        self.balance = 0
        self.returns = self.coins
        self.coins = []
        self._display_balance()
