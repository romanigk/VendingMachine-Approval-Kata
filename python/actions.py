
COINS = {"penny": 1,
         "nickel": 5,
         "dime": 10,
         "quarter": 25}


def select_product(product, machine):
    machine.select_product(product)
    return f"select product: {product}"


def wait_5_secs(machine):
    machine.tick()
    return "wait 5 seconds for display to refresh"


def insert_coin(coin_name, machine):
    machine.insert_coin(COINS[coin_name])
    return f"insert coin: {coin_name}"


def return_coins(machine):
    machine.return_coins()
    return "Return coins"