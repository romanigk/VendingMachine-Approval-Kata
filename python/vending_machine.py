

class VendingMachine:
    def __init__(self, selected_product=None, stock=None):
        self.display = ""
        self.coins = []
        self.balance = 0
        self.selected_product = selected_product
        self.dispensed_product = None
        self.returns = []
        self.bank = []
        self.stock = stock or {}
        self.accepted_coins = [5,10,25]
        self.prices = {"Cola": 100, "Chips": 50, "Candy": 65}
        self._display_balance()

    def _display_balance(self):
        if self.balance:
            self.display = f"{self.balance}"
        elif self._has_change():
            self.display = "INSERT COIN"
        else:
            self.display = "EXACT CHANGE ONLY"

    def insert_coin(self, value):
        if value in self.accepted_coins:
            self.coins.append(value)
            self.balance += value
            self._display_balance()
        else:
            self.returns.append(value)

    def select_product(self, product):
        self.selected_product = product
        if self.balance >= self.prices[self.selected_product]:
            self._dispense_product()
        else:
            cents_to_dollars = '{:,.2f}'
            cost = self.prices[product]/100.0
            self.display = f"PRICE ${cents_to_dollars.format(cost)}"

    def _dispense_product(self):
        if self.stock[self.selected_product] >= 1:
            self.stock[self.selected_product] -= 1
            self.display = "THANK YOU"
            self.dispensed_product = self.selected_product
            self.balance -= self.prices[self.selected_product]
            self.bank += self.coins
            self.coins = []
            if self.balance > 0:
                change = _get_change_required(self.balance)
                self.returns = change
                [self.bank.remove(coin) for coin in change]
                self.balance = 0
            self.selected_product = None

        else:
            self.display = "SOLD OUT"
            self.selected_product = None

    def tick(self):
        if "PRICE" in self.display:
            self._display_balance()
        elif "THANK YOU" in self.display or "SOLD OUT" in self.display:
            self._display_balance()
            self.dispensed_product = None
            self.returns = []
        elif self.selected_product and self.balance >= self.prices[self.selected_product]:
            self._dispense_product()

    def update_stock(self, product, quantity):
        self.stock[product] = quantity

    def update_bank(self, coins):
        self.bank += coins
        self._display_balance()

    def return_coins(self):
        self.balance = 0
        self.returns = self.coins
        self.coins = []
        self._display_balance()

    def _has_change(self):
        quarters = [coin for coin in self.bank if coin == 25]
        dimes = [coin for coin in self.bank if coin == 10]
        nickels = [coin for coin in self.bank if coin == 5]
        return quarters and dimes and nickels


def _get_change_required(balance):
    # TODO: don't fake this one!
    return [10, 5]
