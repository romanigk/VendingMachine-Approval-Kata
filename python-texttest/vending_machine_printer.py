from vending_machine import VendingMachine


class VendingMachinePrinter:
    def __init__(self, machine : VendingMachine):
        self.columns = 60
        self.machine = machine

    def print_everything(self, include_empty=True):
        text = "VendingMachine\n"
        text += self.format_line_with_whitespace("Display", self.machine.display)
        fields = [
            ("Balance", self.machine.balance),
            ("Coins", self.machine.coins),
            ("Returns", self.machine.returns),
            ("Selected product", self.machine.selected_product),
            ("Dispensed product", self.machine.dispensed_product),
            ("Stock", self.machine.stock),
            ("Banked Coins", self.machine.bank),
        ]
        for name, field in fields:
            if field or include_empty:
                text += self.format_line_with_whitespace(name, field)

        text += "\n"
        return text

    def print_significant_fields(self):
        return self.print_everything(include_empty=False)

    def format_line_with_whitespace(self, name, value):
        value_as_str = f"{value}"
        whitespace_size = self.columns - len(name) - len(value_as_str)
        whitespace = ""
        for i in range(whitespace_size):
            whitespace += " "
        line = f"{name}{whitespace}{value_as_str}\n"
        return line
