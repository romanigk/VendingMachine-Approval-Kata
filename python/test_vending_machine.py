import pytest
from approvaltests import verify

from actions import select_product, wait_5_secs, insert_coin, return_coins
from vending_machine import VendingMachine
from vending_machine_printer import VendingMachinePrinter


class Story:
    def __init__(self, printer):
        self.printer = printer
        self._to_verify = ""

    def init(self, name):
        self._to_verify += name
        self._to_verify += "\n\n"

    def arrange(self):
        self._to_verify += self.printer.print_everything()

    def act(self, action):
        self._to_verify += action
        self._to_verify += "\n\n"
        self._to_verify += self.printer.print_significant_fields()

    def __str__(self):
        return self._to_verify


@pytest.fixture()
def machine():
    return VendingMachine(stock={"Cola": 1, "Chips": 1, "Candy": 1})


@pytest.fixture
def printer(machine):
    return VendingMachinePrinter(machine)


@pytest.fixture
def story(printer):
    return Story(printer)


def test_accept_coins(machine: VendingMachine, story: Story):
    story.init("Feature: Accept coins")
    story.arrange()

    story.act(insert_coin("nickel", machine))
    story.act(insert_coin("dime", machine))
    story.act(insert_coin("quarter", machine))

    verify(str(story))


def test_reject_penny(machine: VendingMachine, story: Story):
    story.init("Feature: Reject penny")
    story.arrange()

    story.act(insert_coin("penny", machine))

    verify(str(story))


def test_select_product(machine : VendingMachine, story: Story):
    story.init("Feature: Select product and see price")
    story.arrange()

    story.act(select_product("Cola", machine))
    story.act(wait_5_secs(machine))
    story.act(select_product("Chips", machine))
    story.act(wait_5_secs(machine))

    verify(str(story))



def test_pay_for_product(machine : VendingMachine, story: Story):
    story.init("Feature: Pay for product")
    machine.selected_product="Cola"
    story.arrange()

    story.act(insert_coin("quarter", machine))
    story.act(insert_coin("quarter", machine))
    story.act(insert_coin("quarter", machine))
    story.act(insert_coin("quarter", machine))
    story.act(wait_5_secs(machine))
    story.act(wait_5_secs(machine))

    verify(str(story))


def test_pay_for_chips(machine : VendingMachine, story: Story):
    story.init("Feature: Pay for Chips")
    machine.selected_product="Chips"
    machine.insert_coin(25)
    story.arrange()
    story.act(insert_coin("quarter", machine))

    story.act(wait_5_secs(machine))
    story.act(wait_5_secs(machine))

    verify(str(story))


def test_pay_then_select_product(machine : VendingMachine, story: Story):
    story.init("Feature: Pay first then select product")
    machine.insert_coin(25)
    machine.insert_coin(25)
    machine.insert_coin(10)
    machine.insert_coin(5)
    story.arrange()

    story.act(select_product("Candy", machine))
    story.act(wait_5_secs(machine))

    verify(str(story))


def test_sold_out(machine : VendingMachine, story: Story):
    story.init("Feature: Sold out of Chips")
    machine.insert_coin(25)
    machine.insert_coin(25)
    machine.update_stock("Chips", 0)
    story.arrange()

    story.act(select_product("Chips", machine))
    story.act(wait_5_secs(machine))

    verify(str(story))


def test_return_coins(machine : VendingMachine, story: Story):
    story.init("Feature: Return Coins")
    machine.insert_coin(25)
    machine.insert_coin(10)
    story.arrange()

    story.act(return_coins(machine))

    verify(str(story))


def test_change_available(machine : VendingMachine, story: Story):
    story.init("Feature: Change is available")
    machine.update_bank([25, 25, 25, 25, 10, 10, 10, 10, 5, 5, 5, 5])
    story.arrange()

    verify(str(story))


def test_give_change(machine : VendingMachine, story: Story):
    story.init("Feature: Return change")
    machine.update_bank([25, 10, 5,])
    machine.insert_coin(25)
    machine.insert_coin(25)
    machine.insert_coin(25)
    story.arrange()

    story.act(select_product("Candy", machine))
    story.act(wait_5_secs(machine))

    verify(str(story))
