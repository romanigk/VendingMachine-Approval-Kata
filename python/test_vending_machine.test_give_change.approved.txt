Feature: Return change

VendingMachine
Display                                                   75
Balance                                                   75
Coins                                           [25, 25, 25]
Returns                                                   []
Selected product                                        None
Dispensed product                                       None
Stock                    {'Cola': 1, 'Chips': 1, 'Candy': 1}
Banked Coins                                     [25, 10, 5]

select product: Candy

VendingMachine
Display                                            THANK YOU
Returns                                              [10, 5]
Dispensed product                                      Candy
Stock                    {'Cola': 1, 'Chips': 1, 'Candy': 0}
Banked Coins                                [25, 25, 25, 25]

wait 5 seconds for display to refresh

VendingMachine
Display                                    EXACT CHANGE ONLY
Stock                    {'Cola': 1, 'Chips': 1, 'Candy': 0}
Banked Coins                                [25, 25, 25, 25]

